### PR TITLE
produce separate bundles for oie, classic

### PR DESCRIPTION
**Pull request by aarongranick-okta**
## Description:

adds new "classic" and "oie" bundles for CDN and NPM (ESM and CommonJS). 
(Limiting support to only one engine shaves more than 500K off the widget size!)

3 main bundles:

- default: includes everything
- classic: includes V1 router and AuthJS client with authn api support
- oie: includes V2 router and AuthJS client with IDX api support

both classic and oie bundles will throw an error if `useClassicEngine` option is misconfigured.

small refactor to use mixin and factory pattern from auth-js

## PR Checklist

- [ ] Have you verified the basic functionality for this change?
- [ ] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-489177](https://oktainc.atlassian.net/browse/OKTA-489177)

### CODE REVIEW

File: webpack.common.config.js [hunk 1]
```diff
@@ -11,7 +11,16 @@ var LOCAL_PACKAGES = resolve(__dirname, 'packages/');
...
+module.exports = function({
+  entry,
+  outputFilename,
+  mode = 'development',
+  engine = '',
+  cdn = true
+}) {
+
+  // normalize entry so it is always an array
+  entry = Array.isArray(entry) ? entry : [entry];
```
Comment (lesterchoi-okta) [inline webpack.common.config.js:23]:
```suggestion
  entry = typeof entry === 'string' ? [entry] : entry;
```

Comment (aarongranick-okta) [reply webpack.common.config.js:23]: What is the problem if the entry is an object and not a string?

File: playground/main.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@
 // eslint-disable-next-line @typescript-eslint/no-var-requires
 const signinWidgetOptions = require('../.widgetrc.js'); // commonJS module
 
-// import OktaSignIn from '../src/widget/OktaSignIn';
+// import OktaSignIn from '../src/exports/default';
```
Comment (shuowu-okta) [inline playground/main.ts:6]: what this commented import for?

**Review by shuowu-okta — APPROVED** [review]

File: scripts/verify-registry-install.sh [hunk 3]
```diff
@@ -31,6 +31,9 @@ artifact_version="$(ci-pkginfo -t pkgname)-$(ci-pkginfo -t pkgsemver)"
 git clone https://github.com/okta/samples-js-angular.git test/package/angular-sample
 pushd test/package/angular-sample/custom-login
 
+# BETA WIDGET BRANCH
+git checkout ag-widget-7
```
Comment (jaredperreault-okta) [inline scripts/verify-registry-install.sh:35]: I believe this can be removed

File: scripts/verify-registry-install.sh [hunk 4]
```diff
@@ -31,6 +31,9 @@ artifact_version="$(ci-pkginfo -t pkgname)-$(ci-pkginfo -t pkgsemver)"
...
+
 # sample is setup to use npm
 # broadcast-channel needs to be installed first to guarantee microtime@3.0.0 is used
 if ! npm i broadcast-channel@4.13.0; then
```
Comment (jaredperreault-okta) [inline scripts/verify-registry-install.sh:39]: I believe this also can be removed

File: src/authClient/classic.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  mixinAuthn,
+  PKCETransactionMeta,
+  OAuthStorageManagerInterface,
+  createOAuthStorageManager,
+  OktaAuthOAuthOptions,
+  createOAuthOptionsConstructor,
+  createTransactionManager
+} from '@okta/okta-auth-js';
+
+
```
Comment (jaredperreault-okta) [inline src/authClient/classic.ts:13]: add comments explaining the goal of the mixin pattern used here

File: src/authClient/default.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// Default supports both Authn and IDX pipelines
+import { mixinAuthn } from '@okta/okta-auth-js';
+import { OktaAuth as OktaAuthIdx } from './oie';
+
```
Comment (jaredperreault-okta) [inline src/authClient/default.ts:4]: add comments explaining the goal of the mixin pattern used here

File: src/authClient/oie.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  IdxTransactionMeta,
+  IdxStorageManagerInterface,
+  createIdxStorageManager,
+  createIdxOptionsConstructor,
+  OktaAuthIdxOptions,
+  IdxTransactionManagerInterface,
+  createIdxTransactionManager
+} from '@okta/okta-auth-js';
+
+
```
Comment (jaredperreault-okta) [inline src/authClient/oie.ts:14]: add comments explaining the goal of the mixin pattern used here

File: src/types/authClient.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import { OAuthStorageManagerInterface, OktaAuthConstructor, OktaAuthOAuthInterface, OktaAuthOAuthOptions, PKCETransactionMeta } from "@okta/okta-auth-js";
+
```
Comment (jaredperreault-okta) [inline src/types/authClient.ts:2]: add some comments explaining the use of generics and constructor interfaces

File: test/app/webpack.config.js [hunk 9]
```diff
@@ -18,6 +25,7 @@ const webpackConfig = {
     libraryTarget: 'umd'
   },
   resolve: {
+    // conditionNames: ['import', 'browser'],
```
Comment (jaredperreault-okta) [inline test/app/webpack.config.js:28]: remove?

File: test/package/cjs/package.json [hunk 10]
```diff
@@ -3,7 +3,7 @@
   "version": "1.0.0",
   "description": "",
   "scripts": {
-    "clean": "rimraf dist && rimraf node_modules",
+    "clean": "rm -rf dist && rm -rf node_modules && rm -rf target && rm -f yarn*",
```
Comment (jaredperreault-okta) [inline test/package/cjs/package.json:6]: the use of `rm` may not be safe for Windows users, not sure if that is a concern here

Comment (aarongranick-okta) [reply test/package/cjs/package.json:6]: not much of a concern here. but we need to be able to wipe everything before node_modules installed. so using an NPM module to wipe stuff doesn't make sense.

**Pull request merged**